java/src/main/java/ai/rapids/cudf/MemoryCleaner.java

```diff
@@ -182,65 +190,129 @@ static void setDefaultGpu(int defaultGpuId) {
     defaultGpu = defaultGpuId;
   }
 
-  private static final Thread t = new Thread(() -> {
-    try {
+  private static class CleanerThread extends Thread {
+    private volatile boolean stopFlag = false;
+    private volatile boolean stopped = false;
+
+    CleanerThread() {
+      super("Cleaner Thread");
+    }
+
+    void cleanCollected() {
       int currentGpuId = -1;
       while (true) {
-        CleanerWeakReference next = (CleanerWeakReference)collected.remove(100);
-        if (next != null) {
-          try {
-            if (currentGpuId != defaultGpu) {
-              Cuda.setDevice(defaultGpu);
-              currentGpuId = defaultGpu;
-            }
-          } catch (Throwable t) {
-            log.error("ERROR TRYING TO SET GPU ID TO " + defaultGpu, t);
-          }
-          try {
-            next.clean();
-          } catch (Throwable t) {
-            log.error("CAUGHT EXCEPTION WHILE TRYING TO CLEAN " + next, t);
+        CleanerWeakReference next = null;
+        try {
+          next = (CleanerWeakReference)collected.remove(100);
+        } catch (InterruptedException e)  {
+          // ignore
+        }
+
+        if (next == null) {
+          // queue is empty, stop this loop
+          return;
+        }
+
+        try {
+          if (currentGpuId != defaultGpu) {
+            Cuda.setDevice(defaultGpu);
+            currentGpuId = defaultGpu;
           }
-          all.remove(next.cleaner.id);
+        } catch (Throwable t) {
+          log.error("ERROR TRYING TO SET GPU ID TO " + defaultGpu, t);
+        }
+        try {
+          next.clean();
+        } catch (Throwable t) {
+          log.error("CAUGHT EXCEPTION WHILE TRYING TO CLEAN " + next, t);
+        }
+        all.remove(next.cleaner.id);
+      }
+    }
+
+    @Override
+    public void run() {
+      while(!stopFlag) {
+        cleanCollected();
+      }
+      stopped = true;
```

Comment:
Should this be in a finally block to be sure that it is set properly? Or could we just check the status of the Thread that is running this instead?

Reply:
Done